src/query/executor.rs

```diff
@@ -0,0 +1,162 @@
+        let mut record_id = 0_u32;
+        for share in shares {
+            match a {
+                None => a = Some(share),
+                Some(a_v) => {
+                    let result = ctx
+                        .clone()
+                        .multiply(RecordId::from(record_id), &a_v, &share)
+                        .await
+                        .unwrap();
+                    results.push(result);
+                    record_id += 1;
+                    a = None;
+                }
+            }
+        }
+
+        assert!(a.is_none());
+    }
+
+    results
+}
+
+#[allow(clippy::unused_async)]
+async fn ipa<F: Field, S: Stream<Item = Vec<u8>> + Send + Unpin>(
+    _ctx: SemiHonestContext<'_, F>,
+    _input: S,
+) -> Vec<Replicated<F>> {
+    todo!()
+}
+
+pub fn start_query<S: Stream<Item = Vec<u8>> + Send + Unpin + 'static>(
+    config: QueryConfig,
+    gateway: Gateway,
+    input: S,
+) -> JoinHandle<Box<dyn Result>> {
+    tokio::spawn(async move {
+        // TODO: make it a generic argument for this function
+        let mut rng = StdRng::from_entropy();
+        // Negotiate PRSS first
+        let step = Step::default().narrow(&config.query_type);
+        let prss = negotiate_prss(&gateway, &step, &mut rng).await.unwrap();
+
+        match config.field_type {
+            FieldType::Fp31 => {
+                let ctx = SemiHonestContext::<Fp31>::new(&prss, &gateway);
+                match config.query_type {
+                    #[cfg(any(test, feature = "test-fixture"))]
+                    QueryType::TestMultiply => {
+                        Box::new(test_multiply(ctx, input).await) as Box<dyn Result>
+                    }
+                    QueryType::IPA => Box::new(ipa(ctx, input).await) as Box<dyn Result>,
+                }
+            }
+            FieldType::Fp32BitPrime => {
+                todo!()
+            }
+        }
+    })
+}
+
+#[cfg(all(test, not(feature = "shuttle")))]
+mod tests {
+    use super::*;
+    use crate::secret_sharing::IntoShares;
+    use crate::test_fixture::{Reconstruct, TestWorld};
+    use futures_util::future::join_all;
+    use futures_util::stream;
+
+    #[tokio::test]
+    async fn multiply() {
+        let world = TestWorld::new().await;
+        let contexts = world.contexts::<Fp31>();
+        let a = [Fp31::from(4u128), Fp31::from(5u128)];
+        let b = [Fp31::from(3u128), Fp31::from(6u128)];
+
+        let helper_shares = (a, b).share().map(|(a, b)| {
+            const SIZE: usize = Replicated::<Fp31>::SIZE;
+            let r = a
+                .into_iter()
+                .zip(b)
+                .flat_map(|(a, b)| {
+                    let mut slice = [0_u8; 2 * SIZE];
+                    a.serialize(&mut slice).unwrap();
+                    b.serialize(&mut slice[SIZE..]).unwrap();
+
+                    slice
+                })
+                .collect::<Vec<_>>();
+
+            Box::new(stream::iter(std::iter::once(r)))
```

Comment:
does `r.into_iter()` not work here?

Reply:
maybe it works, but I think it is the same